warehouse/tuf/services.py

```diff
@@ -217,86 +340,120 @@ def init_targets_delegation(self):
+            targets.signed.add_key(
+                RoleType.BINS.value, Key.from_securesystemslib_key(signer.key_dict)
             )
-        )
 
-        # The 'bins' role delegates trust for target files to 'bin-n' roles based on
-        # target file path hash prefixes.
-        delegate_roles_payload[Role.BINS.value] = list()
-        signers = self._key_storage_backend.get(Role.BIN_N.value)
-        key_ids = [signer.key_dict["keyid"] for signer in signers]
+        # Bump version and expiration, and sign and persist updated 'targets'.
+        self._bump_version(targets)
+        self._bump_expiry(targets)
+        self._sign(targets)
+        self._persist(targets)
+
+        targets_meta.append((Targets.type, targets.signed.version))
+
+        # Create new 'bins' role and delegate trust from 'bins' for all target files to
+        # 'bin-n' roles based on file path hash prefixes, a.k.a hash bin delegation.
+        bins = Metadata(Targets())
+        bins.signed.delegations = Delegations(keys={}, roles={})
+        hash_bins = self._get_hash_bins()
         for bin_n_name, bin_n_hash_prefixes in hash_bins.generate():
-            delegate_roles_payload[Role.BINS.value].append(
-                (
-                    DelegatedRole(
-                        bin_n_name,
-                        key_ids,
-                        self._request.registry.settings[
-                            f"tuf.{Role.BIN_N.value}.threshold"
-                        ],
-                        False,
-                        path_hash_prefixes=bin_n_hash_prefixes,
-                    ),
-                    signers,
-                    metadata_repository._set_expiration_for_role(Role.BIN_N.value),
-                )
+            bins.signed.delegations.roles[bin_n_name] = DelegatedRole(
+                name=bin_n_name,
+                keyids=[],
+                threshold=self._request.registry.settings[
+                    f"tuf.{RoleType.BIN_N.value}.threshold"
+                ],
+                terminating=False,
+                path_hash_prefixes=bin_n_hash_prefixes,
             )
-        snapshot_metadata = metadata_repository.delegate_targets_roles(
-            delegate_roles_payload,
-        )
-        self.bump_snapshot(snapshot_metadata)
 
-    def bump_snapshot(self, snapshot_metadata=None):
+            for signer in self._key_storage_backend.get(RoleType.BIN_N.value):
+                bins.signed.add_key(
+                    bin_n_name, Key.from_securesystemslib_key(signer.key_dict)
+                )
+
+            # Create new empty 'bin-n' roles, bump expiration, and sign and persist
+            bin_n = Metadata(Targets())
+            self._bump_expiry(bin_n)
+            self._sign(bin_n)
+            self._persist(bin_n)
+
+            # FIXME: Possible performance gain by updating 'snapshot' right here, to
+            # omit creation of massive list and iterating over all 'bin-n' roles twice.
```

Comment:
This is worthy of generating some numbers to understand the performance. The counter-argument is that we don't want to bump version & sign as we add each bin-n role. Perhaps the right approach is to do the operation in phases so that we can a) add the `MetaFile` as we iterate over bins here b) `_bump_expiry()`, `_bump_version()`, `_sign()`, `_persist()` only once after the loop.